File: 234-palindrome-linked-list/234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool ans = true;
    
    void recur(ListNode* head, ListNode* &temp)
    {
        if(!head->next)return;
        
        head=head->next;
        recur(head, temp);
        
        // cout<<head->val<<" "<<temp->val<<endl;
        
        if(head->val != temp->val)ans=false;
        temp=temp->next;
    }
    
    bool isPalindrome(ListNode* head) 
    {
        if(!head)return true;
        
        recur(head,head);
        
        return ans;
    }
};

File: 234-palindrome-linked-list/234. Palindrome Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

    public void recur(ListNode head, ArrayList<Integer> arr)
    {
        if(head == null)return;

        recur(head.next, arr);
        arr.add(head.val);
    }

    public boolean isPalindrome(ListNode head) {

        ArrayList<Integer> arr = new ArrayList<Integer>();
        ListNode temp = head;

        recur(temp, arr);
        int i=0;
        while(head != null)
        {
            if(head.val != arr.get(i))return false;
            head = head.next;
            i++;
        }

        return true;
    }
}